src/Streamly/Internal/Data/Parse.hs

```diff
@@ -0,0 +1,258 @@
+
+import Streamly.Internal.Data.Strict
+
+-------------------------------------------------------------------------------
+-- Terminating folds
+-------------------------------------------------------------------------------
+--
+-- XXX any/all should be terminating folds when we have a terminating fold
+-- type.
+--
+-- |
+-- >>> S.parse (PR.any (== 0)) $ S.fromList [1,0,1]
+-- > Right True
+--
+{-# INLINABLE any #-}
+any :: Monad m => (a -> Bool) -> Parser m a Bool
+any predicate = Parser step initial (return . Right)
+    where
+    initial = return False
+    step s a = return $
+        if s
+        then Stop 0 True
+        else
+            if predicate a
+            then Stop 0 True
+            else Yield 0 False
+
+-- |
+-- >>> S.parse (PR.all (== 0)) $ S.fromList [1,0,1]
+-- > Right False
+--
+{-# INLINABLE all #-}
+all :: Monad m => (a -> Bool) -> Parser m a Bool
+all predicate = Parser step initial (return . Right)
+    where
+    initial = return True
+    step s a = return $
+        if s
+        then
+            if predicate a
+            then Yield 0 True
+            else Stop 0 False
+        else Stop 0 False
+
+-------------------------------------------------------------------------------
+-- Taking elements
+-------------------------------------------------------------------------------
+--
+-- | Stops after taking exactly @n@ input elements.
+--
+-- * Stops - after @n@ elements.
+-- * Fails - if the stream ends before it can collect @n@ elements.
+--
+-- >>> S.parse (PR.takeExact 4 FL.toList) $ S.fromList [1,0,1]
+-- > Left "takeExact: Expecting exactly 4 elements, got 3"
+--
+-- /Internal/
+--
+{-# INLINABLE takeExact #-}
+takeExact :: Monad m => Int -> Fold m a b -> Parser m a b
+takeExact n (Fold fstep finitial fextract) = Parser step initial extract
+
+    where
+
+    initial = (Tuple' 0) <$> finitial
+
+    step (Tuple' i r) a = do
+        res <- fstep r a
+        let i1 = i + 1
+            s1 = Tuple' i1 res
+        return $ if i1 < n then Skip 0 s1 else Stop 0 s1
+
+    extract (Tuple' i r) = fmap f (fextract r)
+
+        where
+
+        err =
+               "takeExact: Expecting exactly " ++ show n
+            ++ " elements, got " ++ show i
+
+        f x =
+            if n == i
+            then Right x
+            else Left err
+
+-- | Take at least @n@ input elements, but can collect more.
+--
+-- * Stops - never.
+-- * Fails - if the stream end before producing @n@ elements.
+--
+-- >>> S.parse (PR.takeAtLeast 4 FL.toList) $ S.fromList [1,0,1]
+-- > Left "takeAtLeast: Expecting at least 4 elements, got only 3"
+--
+-- >>> S.parse (PR.takeAtLeast 4 FL.toList) $ S.fromList [1,0,1,0,1]
+-- > Right [1,0,1,0,1]
+--
+-- /Internal/
+--
+{-# INLINABLE takeAtLeast #-}
+takeAtLeast :: Monad m => Int -> Fold m a b -> Parser m a b
+takeAtLeast n (Fold fstep finitial fextract) = Parser step initial extract
+
+    where
+
+    initial = (Tuple' 0) <$> finitial
+
+    step (Tuple' i r) a = do
```

Comment:
Perhaps `i` should be strict, though it'll be in NF anyway due to the comparison later.